arbnode/transaction_streamer.go

```diff
@@ -1438,16 +1453,34 @@ func (s *TransactionStreamer) submitEspressoTransactions(ctx context.Context, ig
 			log.Error("failed to get espresso submitted pos", "err", err)
 			return s.config().EspressoTxnsPollingInterval
 		}
-		bytes, _, err := arbos.ParseEspressoMsg(msg.Message)
+		txns, _, err := arbos.ParseEspressoMsg(msg.Message)
 		if err != nil {
 			log.Error("failed to parse espresso message before submitting", "err", err)
 			return s.config().EspressoTxnsPollingInterval
 		}
 
+		if s.dataSigner == nil {
+			panic("data signer not initialized, needs to be initialized to sign the transaction before submitting it to espresso")
+		}
+
+		// Signing the transaction here before sending it to espresso so that external parties have a way to verify that a transaction belongs to a certain namespace
+		// Internally the signature doesn't need to verified because the batch poster receives the transactions from the sequencer.
+		payloadSignature, err := s.dataSigner(crypto.Keccak256Hash(txns[0]).Bytes())
+		if err != nil {
+			log.Error("failed to sign transaction hash before submitting to espresso", "err", err)
+			return s.config().EspressoTxnsPollingInterval
+		}
+
+		var signedPayload []byte
+
+		signedPayload = append(signedPayload, payloadSignature...)
+		signedPayload = append(signedPayload, txns[0]...)
+
 		log.Info("submitting transaction to espresso using sovereign sequencer")
 
+		// Note: same key should not be used for two namespaces for this to work
 		hash, err := s.espressoClient.SubmitTransaction(ctx, espressoTypes.Transaction{
-			Payload:   bytes[0],
+			Payload:   signedPayload,
 			Namespace: s.config().EspressoNamespace,
 		})
```

Comment:
Perhaps we can refactor the logic in the transaction streamer into a separate method, or an external function. This method could take a list of txns, a dataSigner function, and the namespace. It would  build and return the transaction to be passed to the `espressoClient.SubmitTransaction` method. 

This way we could write a unit test over the function rather than having to run a network to observe the results. We could also do this for just the payload itself.

Reply:
IMO, there is still sense in testing the end to end flow because if we create the test centric to this function we will not be able to test the rest of the parts of the PR like changes in `parsing` logic in arbos/parse_l2.go